post-processor/vsphere/post-processor.go

```diff
@@ -169,35 +183,58 @@ func (p *PostProcessor) PostProcess(ctx context.Context, ui packer.Ui, artifact
 
 	ui.Message(fmt.Sprintf("Uploading %s to vSphere", source))
 
-	log.Printf("Starting ovftool with parameters: %s",
-		filterLog(strings.Join(args, " "), ovftool_uri))
-
-	var errWriter io.Writer
-	var errOut bytes.Buffer
-	cmd := exec.Command(ovftool, args...)
-	errWriter = io.MultiWriter(os.Stderr, &errOut)
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = errWriter
+	log.Printf("Starting ovftool with parameters: %s", strings.Join(args, " "))
 
-	if err := cmd.Run(); err != nil {
-		err := fmt.Errorf("Error uploading virtual machine: %s\n%s\n", err, filterLog(errOut.String(), ovftool_uri))
+	ui.Message("Validating Username and Password with dry-run")
+	err = p.ValidateOvfTool(args, ovftool)
+	if err != nil {
 		return nil, false, false, err
 	}
 
-	ui.Message(filterLog(errOut.String(), ovftool_uri))
+	// Validation has passed, so run for real.
+	ui.Message("Calling OVFtool to upload vm")
+	commandAndArgs := []string{ovftool}
+	commandAndArgs = append(commandAndArgs, args...)
+	comm := &shelllocal.Communicator{
+		ExecuteCommand: commandAndArgs,
+	}
+	flattenedCmd := strings.Join(commandAndArgs, " ")
+	cmd := &packer.RemoteCmd{Command: flattenedCmd}
+	log.Printf("[INFO] (vsphere): starting ovftool command: %s", flattenedCmd)
+	if err := cmd.RunWithUi(ctx, comm, ui); err != nil {
+		return nil, false, false, fmt.Errorf(
+			"Error uploading virtual machine: Please see output above for more information.")
+	}
 
 	artifact = NewArtifact(p.config.Datastore, p.config.VMFolder, p.config.VMName, artifact.Files())
 
 	return artifact, false, false, nil
 }
 
-func filterLog(s string, u *url.URL) string {
-	password, passwordSet := u.User.Password()
-	if passwordSet && password != "" {
-		return strings.Replace(s, password, "<password>", -1)
-	}
+func (p *PostProcessor) ValidateOvfTool(args []string, ofvtool string) error {
+	args = append([]string{"--verifyOnly"}, args...)
+	var out bytes.Buffer
+	cmdCtx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	cmd := exec.CommandContext(cmdCtx, ovftool, args...)
+	cmd.Stdout = &out
+
+	// Need to manually close stdin or else the ofvtool call will hang
+	// forever in a situation where the user has provided an invalid
+	// password or username
+	stdin, _ := cmd.StdinPipe()
+	defer stdin.Close()
```

Comment:
in the unlikely case `cmd.StdinPipe()` errors, calling `stdin.Close()` will panic

```suggestion
	stdin, err := cmd.StdinPipe()
	if err != nil {
		err
	}
	defer stdin.Close()
```

Reply:
https://golang.org/src/os/exec/exec.go?s=16663:16712#L567